include/openssl/hpke.h

```diff
@@ -0,0 +1,356 @@
+
+# include <openssl/ssl.h>
+
+/** biggest/default buffer for keys and internal buffers we use */
+# ifndef HPKE_MAXSIZE
+#  define HPKE_MAXSIZE (2 * 1024) /* 2k: enough for anyone :-) */
+# endif
+
+/*
+ * The HPKE modes
+ */
+# define HPKE_MODE_BASE              0 /**< Base mode  */
+# define HPKE_MODE_PSK               1 /**< Pre-shared key mode */
+# define HPKE_MODE_AUTH              2 /**< Authenticated mode */
+# define HPKE_MODE_PSKAUTH           3 /**< PSK+authenticated mode */
+
+/*
+ * The (16bit) HPKE algorithn IDs
+ */
+# define HPKE_KEM_ID_RESERVED         0x0000 /**< not used */
+# define HPKE_KEM_ID_P256             0x0010 /**< NIST P-256 */
+# define HPKE_KEM_ID_P384             0x0011 /**< NIST P-256 */
+# define HPKE_KEM_ID_P521             0x0012 /**< NIST P-521 */
+# define HPKE_KEM_ID_25519            0x0020 /**< Curve25519 */
+# define HPKE_KEM_ID_448              0x0021 /**< Curve448 */
+
+# define HPKE_KDF_ID_RESERVED         0x0000 /**< not used */
+# define HPKE_KDF_ID_HKDF_SHA256      0x0001 /**< HKDF-SHA256 */
+# define HPKE_KDF_ID_HKDF_SHA384      0x0002 /**< HKDF-SHA512 */
+# define HPKE_KDF_ID_HKDF_SHA512      0x0003 /**< HKDF-SHA512 */
+# define HPKE_KDF_ID_MAX              0x0003 /**< HKDF-SHA512 */
+
+# define HPKE_AEAD_ID_RESERVED        0x0000 /**< not used */
+# define HPKE_AEAD_ID_AES_GCM_128     0x0001 /**< AES-GCM-128 */
+# define HPKE_AEAD_ID_AES_GCM_256     0x0002 /**< AES-GCM-256 */
+# define HPKE_AEAD_ID_CHACHA_POLY1305 0x0003 /**< Chacha20-Poly1305 */
+# define HPKE_AEAD_ID_MAX             0x0003 /**< Chacha20-Poly1305 */
+
+/* strings for modes */
+# define HPKE_MODESTR_BASE       "base"    /**< base mode (1), no sender auth */
+# define HPKE_MODESTR_PSK        "psk"     /**< psk mode (2) */
+# define HPKE_MODESTR_AUTH       "auth"    /**< auth (3) with sender-key pair */
+# define HPKE_MODESTR_PSKAUTH    "pskauth" /**< psk+sender-key pair (4) */
+
+/* strings for suites */
+# define HPKE_KEMSTR_P256        "p256"              /**< KEM id 0x10 */
+# define HPKE_KEMSTR_P384        "p384"              /**< KEM id 0x11 */
+# define HPKE_KEMSTR_P521        "p521"              /**< KEM id 0x12 */
+# define HPKE_KEMSTR_X25519      "x25519"            /**< KEM id 0x20 */
+# define HPKE_KEMSTR_X448        "x448"              /**< KEM id 0x21 */
+# define HPKE_KDFSTR_256         "hkdf-sha256"       /**< KDF id 1 */
+# define HPKE_KDFSTR_384         "hkdf-sha384"       /**< KDF id 2 */
+# define HPKE_KDFSTR_512         "hkdf-sha512"       /**< KDF id 3 */
+# define HPKE_AEADSTR_AES128GCM  "aes128gcm"         /**< AEAD id 1 */
+# define HPKE_AEADSTR_AES256GCM  "aes256gcm"         /**< AEAD id 2 */
+# define HPKE_AEADSTR_CP         "chachapoly1305"    /**< AEAD id 3 */
+
+/*
+ * @brief ciphersuite combination
+ */
+typedef struct {
+    uint16_t    kem_id; /**< Key Encryption Method id */
+    uint16_t    kdf_id; /**< Key Derivation Function id */
+    uint16_t    aead_id; /**< AEAD alg id */
+} hpke_suite_t;
```

Comment:
if this stay public, it should follow openssl conventions :
```
typedef struct hpke_suite_t {
    uint16_t    kem_id;  /**< Key Encryption Method id */
    uint16_t    kdf_id;    /**< Key Derivation Function id */
    uint16_t    aead_id;  /**< AEAD alg id */
} OSSL_HPKE_SUITE ;
```

Reply:
`struct ossl_hpke_suite_st`?

Reply:
Fair. I've renamed more or less everything externally visible starting with ``hpke_`` or ``HPKE_`` to be ``ossl_hpke_`` or ``OSSL_HPKE_`` and ``hpke_suite_t`` to ``ossl_hpke_suiite_st`` - need to do a bit of testing (tomorrow, my time) and will then push a version that should address all of the fine @tmshort and @FdaSilvaYY comments above - will submit a cumulative comment then too. (Meanwhile, many thanks for all the fine comments.)

Reply:
The struct name isn't mandatory in this case